Group instanced pipelines on UI

aoldershaw opened the pull request:
<!--
Hi there! Thanks for submitting a pull request to Concourse!

The title of your pull request will be used to generate the release notes.
Please provide a brief sentence that describes the PR, using the [imperative
mood]. Please refrain from adding prefixes like 'feature:', and don't include a
period at the end.

Examples: "Add feature to doohickey", "Fix panic during spline reticulation"

We will edit the title if needed so don't worry about getting it perfect!

To help us review your PR, please fill in the following information.
-->

[imperative mood]: https://chris.beams.io/posts/git-commit/#imperative

## What does this PR accomplish?
<!--
Choose all that apply.
Also, mention the linked issue here.
This will magically close the issue once the PR is merged.
-->
Bug Fix | **Feature** | Documentation

closes #5921 .
closes #6391 .

## Changes proposed by this PR:
<!--
Tell the reviewer What changed, Why, and How were you able to accomplish that?
-->

* ~Changed all pipeline-related routes in the UI from `/teams/:team_name/pipelines/:pipeline_name` to `/pipelines/:pipeline_id`~
  * Nevermind - I ended up changing it back
  * All pipeline routes also support instance vars in the query string in the form `var.some_instance_var="value"` (`some_instance_var` can use dot notation if it represents nested JSON)
* Render instance group cards on the dashboard
  * This card displays the status of all pipelines with the same name (but different instance vars). Pipelines are displayed in a grid, similarly to how jobs are displayed on pipeline cards (but without any meaningful structure)
  * The card has a badge indicating the number of pipelines in the group
  * Clicking on a pipeline square in the group takes you directly to the pipeline page
  * Clicking on the card header takes you to the "instance group view"
* Add a "new" instance group view
  * Displays all instances within an instance group as modified pipeline cards
    * These cards have a header that displays the instance vars in a flattened JSON notation, rather than displaying the pipeline name
    * e.g. `{"version": {"name": "latest", "semver": "1.2.3"}}` is rendered as `version.name: latest`, `version.semver: 1.2.3` (on separate lines)
  * This isn't so much a new view, and more of an extension of the dashboard
  * The intention in the future is that you will be able to navigate to this view by the filters that are currently applied - i.e. if you are filtering specifically for instance group(s), you will see this view
  * This PR doesn't add drag and drop functionality on the instance group view
* On either dashboard or instance group view, you can filter instances by fuzzy searching
  * You can fuzzy search by the values of instance vars (putting spaces between each instance var value)
* When pipeline instances are created, they are assigned the same team-level ordering as existing instances in the same instance group
* Pipeline/job/build/resource pages for instances have the instance group name in the breadcrumbs, and the pipeline name is replaced by a simplified notation of the instance vars
  * This is the "hyphen notation" - we take all of the values of instance vars, and join them to a `-`
  * This removes all instance var keys
  * We also use this notation for pipeline tooltips in the instance group view

## Notes to reviewer:
<!--
Leave a message to whoever is going to review this PR.
Mainly, pointers to review the PR, and how they can test it.
-->

Firstly, I'm sorry for the huge diff 😬 I would *definitely* recommend reviewing this PR commit by commit. There's a pretty large set of changes up front with the renaming of routes to `/pipelines/:pipeline_id` - after that, I tried to be pretty focused with commits, and left pretty thorough commit messages.

~Also, this is currently blocked on #6085 (hence the failing watsjs/upgrade/downgrade) - if you want to test this PR out, I'd recommend building the `concourse/concourse:local` image from there (i.e. checkout that branch, `docker-compose up --build -d`, switch back to this branch, and then `yarn build`). Unfortunately, this means you won't build the changes to the pipeline ordering API, so drag and drop on the dashboard won't work - but once that's merged, you can test that out too.~

In order to test this, you'll need to set some pipeline instances! You can do that by running e.g.:

```
$ fly -t dev set-pipeline -p release -c pipeline.yml -i version=6.5.0
```

The `-i` flag is new, and is used for instance vars. You can also set nested JSON using dot notation, e.g.:

```
... -i foo.bar=123 -i foo.baz=abc
# produces {"foo": {"bar": 123, "baz": "abc"}}
```

Unrelated to this PR, but if you want to reference these pipelines via `fly`, you can run:

```
$ fly -t dev get-pipeline -p release/version:6.5.0
# or for nested JSON
$ fly -t dev get-job -j release/foo.bar:123,foo.baz:abc/some-job
# or alternatively
$ fly -t dev get-job -j 'release/foo:{bar:123,baz:abc}/some-job'
```

## Release Note
<!--
If needed, you can leave a detailed description here which will be used to
generate the release note for the next version of Concourse. The title of the
PR will also be pulled into the release note.
-->

TODO

## Contributor Checklist
<!--
Most of the PRs should have the following added to them,
this doesn't apply to all PRs, so it is helpful to tell us what you did.
-->
- [ ] Followed [Code of conduct], [Contributing Guide] & avoided [Anti-patterns]
- [ ] [Signed] all commits
- [ ] Added tests (Unit and/or Integration)
- [ ] Updated [Documentation]
- [ ] Added release note (Optional)

[Code of Conduct]: https://github.com/concourse/concourse/blob/master/CODE_OF_CONDUCT.md
[Contributing Guide]: https://github.com/concourse/concourse/blob/master/CONTRIBUTING.md
[Anti-patterns]: https://github.com/concourse/concourse/wiki/Anti-Patterns
[Signed]: https://help.github.com/en/github/authenticating-to-github/signing-commits
[Documentation]: https://github.com/concourse/docs

## Reviewer Checklist
<!--
This section is intended for the reviewers only, to track review
progress.
-->
- [ ] Code reviewed
- [x] Tests reviewed
- [ ] Documentation reviewed
- [ ] Release notes reviewed
- [ ] PR acceptance performed
~~- [ ] New config flags added? Ensure that they are added to the
  [BOSH](https://github.com/concourse/concourse-bosh-release) and
  [Helm](https://github.com/concourse/helm) packaging; otherwise, ignored for
  the [integration
  tests](https://github.com/concourse/ci/tree/master/tasks/scripts/check-distribution-env)
  (for example, if they are Garden configs that are not displayed in the
  `--help` text).~~

xtreme-vikram-yadav commented on web/elm/src/Api/Endpoints.elm:
```diff
@@ -118,10 +118,10 @@ toPath endpoint =
 
         Job id subEndpoint ->
-            pipelinePath id ++ [ "jobs", id.jobName ] ++ jobEndpointToPath subEndpoint
+            pipelinePath id.pipelineId ++ [ "jobs", id.jobName ] ++ jobEndpointToPath subEndpoint
```
Seems like an existing convention but it reads weird to access properties on id

aoldershaw replied:
Yeah, perhaps a bit weird to read. The `Identifier` types aren't (necessarily) the same thing as a database ID, though - they're just some way to uniquely identify entities. The exception is now `PipelineIdentifier`, which is uniquely identified by the database ID - but e.g. `JobIdentifier` is identified by the `pipelineId` and `jobName`, since that's what it takes to uniquely specify a job

zoetian replied:
It is really weird actually :p but we can probably improve it in future

aoldershaw commented:
@zoetian @xtreme-vikram-yadav heads up, I'm working on reverting the first 4 commits, and keeping the pipeline-related routes as `/teams/:team_name/pipelines/:pipeline_name` - so, I wouldn't get too caught up in the first few commits! Sorry if you've already sunk time into that 😬

aoldershaw commented:
@xtreme-vikram-yadav and @zoetian , this is now ready for review and no longer dependent on #6085, so you can `docker-compose up` from this branch

aoldershaw commented:
Hey @zoetian and @xtreme-vikram-yadav - would love to get some feedback on this when you get the chance! I recognize it's a large change, but I'm happy to walk through it with you if you'd like

zoetian commented:
Thaaaanks @aoldershaw! I'm slowly working on reviewing it. Should be done by the end of this week... 🤞

zoetian commented on web/elm/tests/Data.elm:
```diff
@@ -261,6 +311,37 @@ withBuildName name p =
+
+withId : Int -> { r | id : Int } -> { r | id : Int }
+withId id j =
+    { j | id = id }
```
👍  these helpers in test are really neat!

zoetian commented on web/elm/tests/SideBarFeature.elm:
```diff
@@ -1083,10 +1083,7 @@ iSeeItHasAValidPipelineId =
-                ++ Base64.encode "team"
-                ++ "_"
-                ++ Base64.encode "pipeline"
+                ++ "_1"
```
intéressant.. guess we have to keep the underscore here 🤔

zoetian commented:
I just started testing the feature manually, as a small break from code review... should these div banners also change the color by the pipeline jobs' states?

<img width="1215" alt="Screen Shot 2020-11-06 at 10 21 11 AM" src="https://user-images.githubusercontent.com/6901732/98383652-c61d9c80-201a-11eb-8742-5155354da077.png">

aoldershaw commented:
@zoetian we decided that based on the nature of instanced pipelines, an overall "status" of the instance group isn't that telling. e.g. if you have an instance group for PRs, it's not that interesting to see that at least one of the PRs failed. So, we just went with an arbitrary off-grey to match the shape of other pipeline cards

zoetian commented on web/elm/src/Dashboard/Dashboard.elm:
```diff
@@ -1161,29 +1173,36 @@ pipelinesView session params =
+
+        cardsByTeam =
+            filteredPipelinesByTeam
+                |> List.map (\( team, ps ) -> ( team, ps |> List.map PipelineCard ))
```
ugh... I'm not sure if I understand the two maps here tbh...
So the `filteredPipelinesByTeams` returns all the pipelines and pass them into the map, where the actual function `(\( team, ps ) -> ( team, ps |> List.map PipelineCard ))` is a mapping from a tuple of `(team, ps)` to the  function`PipelineCard`?

I'm confused about why we need this tuple here... it seems like PipelineCard only needs a pipeline?

aoldershaw replied:
Maybe outlining the types explicitly would make it more clear.

We have

```
filteredPipelinesByTeam : List (String, List Pipeline)
```

and we want

```
cardsByTeam : List (String, List Card)
```

so the first `map` converts the outer `List` (of tuples), and the second `map` converts the inner `List` within each tuple (of `Pipelines`). The point of the tuple is just so we can render each team's cards individually

aoldershaw commented:
FYI @clarafu and @chenbh, while this is ready for review, it's now dependent on backend changes in #6177 - so, you'll have to build concourse from that branch until it's merged. Unfortunately, that means the backend changes in this PR (the reorder pipelines API) won't work, so reordering cards on the dashboard will be borked.

clarafu commented:
I took a skim over the code within ATC and it looked good to me! For the web UI changes, I'm just going to do acceptance for that because I probably wouldn't be able to give much insight into the elm code changes.

Just to make sure I am building this correctly when doing my acceptance testing, I ran `yarn build` in the `issue/5921` branch and then did `git co instanced-pipeline-fixup` to switch to the backend changes and did `docker-compose up --build`. 

One thing that I found which might not need to be included in this PR is favouriting of instances. I saw that you wrote in this [comment](https://github.com/concourse/concourse/issues/5921#issuecomment-692110998) that you can favourite individual instances and they would show up in the "preview" of the instance group. When I tried it out locally, it seems like all the instances still show up when I click on a favourited instance in the side bar. But I'm not sure if you were going to implement it in a separate PR or was your comment just outdated!

Also I created https://github.com/concourse/concourse/issues/6278 because I created an instance through `fly set-pipeline` with a borked var value and was trying to delete it but found out there was no way to do that haha. 

I haven't finished doing acceptance yet but these are just the notes I wrote down so far!

aoldershaw commented:
> Just to make sure I am building this correctly when doing my acceptance testing, I ran yarn build in the issue/5921 branch and then did git co instanced-pipeline-fixup to switch to the backend changes and did docker-compose up --build.

👍 

> I saw that you wrote in this comment that you can favourite individual instances and they would show up in the "preview" of the instance group. When I tried it out locally, it seems like all the instances still show up when I click on a favourited instance in the side bar. But I'm not sure if you were going to implement it in a separate PR or was your comment just outdated!

The main benefit to favouriting instances is that the instance group card that shows up in the favourite pipelines section on the dashboard only displays the favourited pipelines, making it easier to navigate directly to the pipeline page rather than having to go through the instance group page first.

With the sidebar, it's not as useful - you're effectively only able to see instance groups that have a favourited pipeline. Perhaps if there's just a single favourited pipeline in an instance group, the sidebar link could take you directly to that pipeline?

I talked to @matthewpereira about adding a favourites section to the instance group view (like we have on the regular dashboard), but we decided against it, as there may be a better paradigm for displaying favourited pipelines (possible with the view selector idea he had showcased a little while back).

Perhaps as an alternative, we could add a `favorites: true` filter (or something like that) for the search bar, and the sidebar link could also apply that filter. But if we went with that approach, I'd leave it for a subsequent PR.

> Also I created #6278 because I created an instance through fly set-pipeline with a borked var value and was trying to delete it but found out there was no way to do that haha.

Commented on that issue!

clarafu commented:
> The main benefit to favouriting instances is that the instance group card that shows up in the favourite pipelines section on the dashboard only displays the favourited pipelines, making it easier to navigate directly to the pipeline page rather than having to go through the instance group page first.

Ah okay, I thought what you meant by your [comment](https://github.com/concourse/concourse/issues/5921#issuecomment-692110998) was that when you favourite an instance(s) and you click on the favourited pipeline (through the sidebar probably), it would only "preview" the favourited instance(s) in the instances group view page. 

I do think it is useful for favourited instances to be show in the favourite pipelines in the dashboard, but I also think it is a little confusing for it to jump to the instances group view with all the instances (favourited or not) when you click on a favourited instance in the sidebar. I think the part that makes it confusing is because in the sidebar it shows the number of instances in that group, and when you favourite an instance it will show `1` for the number of instances in the instance group within the favourited section. But then when you click on it, it shows the instances group view with all the instances including the non favourited ones. I think it might make more sense if we either had the favourited instance group in the sidebar always show the same number of instances (regardless of favourited or non favourited) so that when you click on the instance group in the sidebar (either in the favourite section or not), it will always jump you to the instances group view with all the instances. Or another option is for the sidebar to show the count of favourited instances and then when you click on it, it will only display the favourited instances within the instances group view page. Also it would be nice if there is only one favourited instance, it directs you straight to the pipeline view of that instance. That being said, this is just my opinion though and I can definitely be easily persuaded to something else that makes sense!

aoldershaw commented:
@clarafu Taylor and I discussed with Matthew, and we decided that it might be less jarring/easier to navigate if we just also included a favorite pipelines section in the instance group view - that way, you at least see the pipelines you care about at the top of the screen.

We still need to figure out the best way to approach favouriting wrt instance groups, but think we're going to wait to get some more broad feedback before changing much else.

Let me know if you think this is an improvement or whether you think it was better before!

clarafu commented:
I'm still working on acceptance, but Aidan and I had a discussion on slack about how it might be nice to have a message within the `fly -t <target> set-pipeline` command that will show what pipeline name and instance vars you are trying to set. For example, the output of `fly set-pipeline` would include something like:

```
$ fly -t dev sp -p test -c ... -i foo=bar,baz=baz
... (diff)
pipeline name: test
pipeline instance vars:
  foo: bar,baz=baz
apply configuration? [yN]:
```

or if there is nested vars then it will show it in the YAML format

```
$ fly -t dev sp -p test -c ... -i foo.bar.baz=123
... (diff)
pipeline name: test
pipeline instance vars:
  foo:
    bar:
      baz: 123
apply configuration? [yN]:
```

If there are no changes to apply then it will not show. I'm not sure if this addition will be part of this story or a new one but I just wanted to document it somewhere for now!

I've been continuing on the acceptance and so far there are a few things that I noticed, but again I'm unsure if I'm just building it wrong or is it really a bug because I completely do not know how to read the elm code to figure it out for myself 😅   One thing I noticed was the icons at the top bar where the pipeline name should appear seems to be missing for non instanced pipelines

![Screen Shot 2020-11-19 at 11 55 51 AM](https://user-images.githubusercontent.com/19290703/99698310-9f218a80-2a5e-11eb-8da7-89f2ea7674ae.png)

This is the icon that I would expect to see for regular pipelines.
![Screen Shot 2020-11-19 at 11 56 36 AM](https://user-images.githubusercontent.com/19290703/99698317-a183e480-2a5e-11eb-882b-e073dc973d5d.png)

Also there is a large spacing between the slash and the job name for a job within an instanced pipeline

![Screen Shot 2020-11-19 at 12 00 09 PM](https://user-images.githubusercontent.com/19290703/99698551-ea3b9d80-2a5e-11eb-8959-97cf32970121.png)

I'm guessing that is because that is where the `<>` was for jobs within non instanced pipelines?

![Screen Shot 2020-11-19 at 12 01 45 PM](https://user-images.githubusercontent.com/19290703/99698694-12c39780-2a5f-11eb-9af9-96f10566209e.png)

Another thing I wanted to ask about the top bar is that I was wondering if we were still going to have the top header show `<instance group name>-<instance var 1>[-<instance var 2>][...]` like mentioned in #5921? Or did we change it to just be `<instance var 1>[-<instance var 2>][...]`?

One last thing I noticed was when I clicked on a job from a instanced pipeline view, it would direct me to `http://localhost:8080/teams/main/pipelines/test/jobs/trigger-job/builds/28` rather than `http://localhost:8080/teams/main/pipelines/test/jobs/trigger-job/builds/31?vars.version=%221.2.3%22`.

![demo](https://media.giphy.com/media/LxjLQ64YDdgheklrtg/giphy.gif)

Honestly I'm not confident that I built it correctly so please check to see if you also see the same things as me or is it just because I built it incorrectly that I'm seeing these bugs!

aoldershaw commented:
> I'm not sure if this addition will be part of this story or a new one but I just wanted to document it somewhere for now!

I opened #6300 that does this. Figured it would be better as a separate PR

---

With respect to your other comments, I *think* they might be solved by doing a hard refresh on the pipeline page (shift-click on refresh icon in Chrome at least). Also, you'll need to make sure you're on this branch when you do the hard refresh, because we serve the assets from your local filesystem.

clarafu commented:
@aoldershaw AH that is where my inexperience in web UI comes into play. When I did all the refreshes I was always on the `instanced-pipeline-fixup` branch 😅 . Okay I think all the bugs I saw can be ignored! But I did have the question of the formatting of the instance vars in the top bar:

> Another thing I wanted to ask about the top bar is that I was wondering if we were still going to have the top header show <instance group name>-<instance var 1>[-<instance var 2>][...] like mentioned in #5921? Or did we change it to just be <instance var 1>[-<instance var 2>][...]?

aoldershaw commented:
Ah sorry missed that question. IMO it's a bit redundant to prefix the instance vars with the pipeline name given that the pipeline name is right next to it - feel free to disagree, though.

> Oh I also have a question that is unrelated, is there a way to rename a pipeline that is instanced? So that if you rename a pipeline, it will rename all the it's pipeline instances? I think right now it will only rename a pipeline instance that has no instance vars.

Ah, good point - I hadn't really thought of that. Perhaps `fly rename-pipeline` could work the same way `fly order-pipelines` works as of 0e6869892324b7921e2ca4fe9b440dbbf506b778 (i.e. it takes a pipeline name rather than a pipeline ref)? Though, that would mean you can't move a pipeline instance between instance groups - not sure if that's a common use-case, though.

> Hm I'm also wondering if the instance group name should be in the top header when you are viewing the instance group view? So in the example below, dev should be test instead?

This was a deliberate decision after some discussion with @matthewpereira a while back, but I don't remember 100% why (removed in e08105e4d97f63e4d8b5abd899110611772662c7). I think it's at least in part related to https://github.com/concourse/concourse/issues/5921#issuecomment-691680523 - if viewing an instance group is a special case of filtering, we're still conceptually on the dashboard. That said, it's not obvious currently that it is a special case of filtering yet, so idk

chenbh commented on web/elm/src/Application/Application.elm:
```diff
@@ -426,20 +399,26 @@ urlUpdate : Routes.Route -> Model -> ( Model, List Effect )
+            model.session
+
+        newSession =
+            { oldSession | route = route }
```
I **think** this is the place where we need to clear out the `.hovered` state

Now that the HoverState is tracked in the model, we'll need to clear it every time we navigate to a new page. Otherwise if you hover over a a job in regular pipeline card, clicking into it, and then navigating back to the dashboard page, you'll get a dangling hover text:

![Screen Shot 2020-11-20 at 4 26 01 PM](https://user-images.githubusercontent.com/17577073/99851467-39f89280-2b4d-11eb-9de2-c2e5ea8dfd94.png)
![Screen Shot 2020-11-20 at 4 26 11 PM](https://user-images.githubusercontent.com/17577073/99851469-3a912900-2b4d-11eb-9d0c-ffca61154171.png)

chenbh replied:
Also, is it just me or is the `no instance vars` text slightly higher than the `foo:..` text?

matthewpereira replied:
+1 on the line height (or padding/margin?) for `no instance vars` above. Nice catch, @chenbh !

aoldershaw replied:
Two good catches, thanks! Fixed

chenbh reviewed (commented):
Still working through it, here's some very early thoughts

aoldershaw commented on web/elm/src/SideBar/SideBar.elm:
```diff
@@ -230,6 +228,8 @@ view model currentPipeline =
         in
         Html.div
             (id "side-bar" :: Styles.sideBar newState)
+            -- I'd love to use the curPipeline function instead of passing it in to view,
+            -- but that doesn't work for OneOffBuilds that point to a JobBuild
```
The `OneOffBuild` page isn't so well named tbh. It's for the route `/builds/:build_id`, which could be for a true one-off build, but also could be a regular job build (e.g. when you click "Details" on the GitHub `unit` check, it takes you to `/builds/:build_id`, but it's still just a job build for job `prs/unit`).

In this case, we treat it like the job build page (by highlighting `prs` in the sidebar). But for true one-off builds, we don't highlight any pipeline

aoldershaw commented on web/elm/src/Concourse.elm:
```diff
@@ -813,15 +895,47 @@ type alias PipelineName =
+    }
+
+
+toPipelineId : { r | teamName : TeamName, name : PipelineName, instanceVars : InstanceVars } -> PipelineIdentifier
+toPipelineId p =
+    { teamName = p.teamName
+    , pipelineName = p.name
+    , pipelineInstanceVars = p.instanceVars
```
Yeah....it's not the nicest, the difference in field names is kind of annoying

aoldershaw commented on web/elm/src/Colors.elm:
```diff
@@ -268,6 +270,11 @@ paused =
 
+pausedFaded : String
+pausedFaded =
+    "#2776ab"
```
Good call. Will leave this for another PR, though - #6168 didn't change any of the pipeline status colours, so we can do them all in one go

clarafu commented:
> Ah, good point - I hadn't really thought of that. Perhaps `fly rename-pipeline` could work the same way `fly order-pipelines` works as of [0e68698](https://github.com/concourse/concourse/commit/0e6869892324b7921e2ca4fe9b440dbbf506b778) (i.e. it takes a pipeline name rather than a pipeline ref)? Though, that would mean you can't move a pipeline instance between instance groups - not sure if that's a common use-case, though.

Ya I think it would make sense for it to work the same way as `fly order-pipelines` with using a pipeline name. I thought about it not being able to move a pipeline instance between groups but I thought that it might be weird being able to do that anyways because the instance vars should be tied to things set in the pipeline, and if you want to change those var values then it should produce a whole new instance instead of keeping the history of the builds produced by the old vars.

> This was a deliberate decision after some discussion with @matthewpereira a while back, but I don't remember 100% why (removed in [e08105e](https://github.com/concourse/concourse/commit/e08105e4d97f63e4d8b5abd899110611772662c7)). I think it's at least in part related to [#5921 (comment)](https://github.com/concourse/concourse/issues/5921#issuecomment-691680523) - if viewing an instance group is a special case of filtering, we're still conceptually on the dashboard. That said, it's not obvious currently that it is a special case of filtering yet, so idk

I see, I think I wanted to make that suggestion because when I was playing around with instance groups I found myself getting lost sometimes. Usually it was when I was clicking a lot between the dashboard and instance groups, since the pages are widely the same and there is nothing that is indicating whether I am looking at the full dashboard or just the instance group view (other than the fact that it only shows the instances within that group). If we don't want to display the instance group name in the top header, maybe it could be within the filter search bar? That way there is indication that this is just a filtered search of the dashboard? This is just one option though, I would also be fine keeping it the way it is now too. 😁

aoldershaw commented:
> If we don't want to display the instance group name in the top header, maybe it could be within the filter search bar? That way there is indication that this is just a filtered search of the dashboard?

Yup, the plan is for clicking the instance group to apply a filter like this:

![review](https://user-images.githubusercontent.com/26583442/100246791-966e0000-2f07-11eb-8d22-a622077b00a8.png)

which implies improvements to filtering (chip creation + multiple filters), which is extra work that I think is better left for a follow up PR. p.s. that's just a crappy mockup I made quickly, but Matthew made a nicer one somewhere on Figma I believe

chenbh commented on web/elm/src/Dashboard/Group.elm:
```diff
@@ -178,56 +212,71 @@ viewFavoritePipelines session params =
-    , pipelinesWithResourceErrors : Set ( String, String )
-    , pipelineJobs : Dict ( String, String ) (List Concourse.JobIdentifier)
-    , jobs : Dict ( String, String, String ) Concourse.Job
+    , pipelinesWithResourceErrors : Set Concourse.DatabaseID
+    , pipelineJobs : Dict Concourse.DatabaseID (List Concourse.JobName)
+    , jobs : Dict ( Concourse.DatabaseID, Concourse.JobName ) Concourse.Job
+    , dashboardView : Routes.DashboardView
+    , query : String
     }
     -> { a | userState : UserState }
-    -> Group
+    -> ( Concourse.TeamName, List Card )
     -> List (Html Message)
-hdView { pipelineRunningKeyframes, pipelinesWithResourceErrors, pipelineJobs, jobs } session g =
+hdView { pipelineRunningKeyframes, pipelinesWithResourceErrors, pipelineJobs, jobs, dashboardView, query } session ( teamName, cards ) =
```
Now that we pass in the `session`, we can probably grab the `pipelineRunningKeyframes` from there, let's move it to the second argument it'll make it slightly cleaner https://github.com/concourse/concourse/blob/94af3c750b0ccc28931054605a81fc0972e6c2e3/web/elm/src/Dashboard/Dashboard.elm#L1423-L1432

chenbh reviewed (commented):
Instead of a single big review, I think I'm just going to do daily incremental reviews by file. Feel free to hold off addressing the review comments until I'm done.

chenbh commented on web/elm/src/SideBar/SideBar.elm:
```diff
@@ -266,6 +275,16 @@ tooltip { hovered } =
                 , arrow = Just { size = Styles.tooltipArrowSize, color = Colors.tooltipBackground }
                 }
 
+        HoverState.Tooltip (SideBarInstanceGroup _ _ name) _ ->
```
The tooltips for instance groups doesn't show up when you're in the Pipeline, Job, or Resource page. I suspect it's to do with with the fact `Pipeline.tooltip` is an nop, but not sure why the regular pipeline hovers work but not instance pipeline ones https://github.com/concourse/concourse/blob/94af3c750b0ccc28931054605a81fc0972e6c2e3/web/elm/src/Pipeline/Pipeline.elm#L436-L438

I can't figure out how to screenshot with cursor, but I'm hovering over `pipelin...` and there's no tooltip
![Screen Shot 2020-12-03 at 5 15 54 PM](https://user-images.githubusercontent.com/17577073/101095013-4c31f200-358b-11eb-83de-8a79a6965919.png)

However, hovering over regular pipelines work just fine
![Screen Shot 2020-12-03 at 5 15 21 PM](https://user-images.githubusercontent.com/17577073/101095028-4fc57900-358b-11eb-93da-0cd8a9f914a7.png)

aoldershaw replied:
Good catch, fixed in 39c7c01e12fa04de5351f3240e9a447cb0b9b387

chenbh reviewed (commented):
> Instead of a single big review, I think I'm just going to do daily incremental reviews by file.

Done reviewing for now, was able to test everything other than the reordering. Really liked what you did with the hover tooltips, much nicer to test everything natively in elm.

aoldershaw commented on web/elm/src/Dashboard/Filter.elm:
```diff
@@ -91,36 +93,32 @@ runFilter jobs existingJobs f =
     in
     case pf of
+        FuzzyName term ->
+            Simple.Fuzzy.match term (pipeline.name ++ hyphenNotation pipeline.instanceVars)
```
meh, I'm inclined to ignore that case for now until someone brings it up as an issue

aoldershaw commented on web/elm/src/Message/Effects.elm:
```diff
@@ -683,22 +669,72 @@ toHtmlID domId =
+                ++ Base64.encode groupName
+                ++ "_name"
+
+        PipelineCardInstanceVar section p varName _ ->
+            pipelinesSectionName section
+                ++ "_"
+                ++ encodePipelineId p
+                ++ "_var_"
+                ++ Base64.encode varName
```
while it's true that the tooltip text can technically not be unique, I suspect that won't matter most of the time. I'm guessing most people will use the same instance var names across all instances in a single group, and including the var names will likely just add noise/make it harder to quickly identify what pipeline you're looking at IMO

chenbh replied:
I should've clarified that this is the tooltip when hovering over the instance vars while in the instance group view. If I have a really long var name and hover over it, it'll only show the value
![Screen Shot 2021-01-06 at 10 47 46 AM](https://user-images.githubusercontent.com/17577073/103807747-e66ed500-5024-11eb-800a-a0c272ddcc12.png)

aoldershaw replied:
ahh sorry, thanks for the clarification! yeah, makes sense, will fix soon

aoldershaw replied:
updated

aoldershaw commented on web/elm/tests/RoutesTests.elm:
```diff
@@ -68,7 +166,31 @@ all =
+                                }
+                            )
+                        )
+        , test "dashboard hd view ignores search and instance group query params" <|
```
It's mainly for simplicity of implementation, but also because it's hard to come up with a nice way to visualize them - the  `hyphen-notation` we use in tooltips could work, but it's not ideal (and as you note, not unique in all cases)

chenbh commented:
If you don't mind, can you sneak in a quick fix for something that's currently broken on master, the step tooltip text has the same colour as the background:
![Screen Shot 2021-01-04 at 4 33 24 PM](https://user-images.githubusercontent.com/17577073/103581786-a8e63c80-4eaa-11eb-8605-5e05db086cb9.png)

Adding `style "color" Colors.tooltipText` should fix it
https://github.com/concourse/concourse/blob/39c7c01e12fa04de5351f3240e9a447cb0b9b387/web/elm/src/Build/Styles.elm#L238-L257

aoldershaw commented on web/elm/src/Dashboard/Filter.elm:
```diff
@@ -162,8 +157,8 @@ type GroupFilter
-    = Status StatusFilter
-    | FuzzyName String
+    = FuzzyName String
+    | Status StatusFilter
```
lol I guess? not sure why I preferred it this way at the time

aoldershaw commented:
@chenbh I've addressed (or at least acknowledged) all of your concerns. I think the most controversial thing currently is the tooltip using `hyphen-notation` of just the values and no keys. IMO adding keys would add more visual noise, particularly in what we expect to be common cases (where instances have the same "shape" of instance vars)

Happy to talk more about this if you disagree, though!

chenbh commented:
@aoldershaw there seem to be some edge case regarding dragging past the last possible position

https://user-images.githubusercontent.com/17577073/103668232-e8f9fd80-4f44-11eb-932b-3bb6c6599a6d.mov

chenbh commented on atc/api/pipelineserver/order.go:
```diff
@@ -32,10 +31,10 @@ func (s *Server) OrderPipelines(w http.ResponseWriter, r *http.Request) {
-			"pipeline-refs": pipelinesRefs,
+			"pipeline_names": pipelinesNames,
 		})
 		w.WriteHeader(http.StatusInternalServerError)
```
I tried to order a pipeline that doesn't exist, and it gave a very unhelpful message (note the empty Body)
```
failed to order pipelines: Unexpected Response
Status: 500 Internal Server Error
Body:

```

It would be nice to get either a 404 or some sort of indication that i mistyped a pipeline name

aoldershaw replied:
Updated to return a 400 error (not 404, which is intended to mean "The server has not found anything matching the Request-URI" - but this isn't a Request URI issue, it's in the request body). There's context in the response body, at least

aoldershaw commented:
@chenbh okay I guess my `but it works! I swear!` remark was a bit premature 😀 

Fixed in 7072cf5799b8508639316606720e53a486b40542. Also fixed the tooltips (there were some others that I had missed as well), and now give a useful error when reordering a missing pipeline

chenbh reviewed (approved):
I think this is good to go! 

- Went through a sample pipeline and all the resource/job/build/pipeline links work as expected. 
- Favouriting instance pipelines can only be done from within the instance group, and it'll create a filtered instance group in the favourites section 
- Reordering pipelines intentionally does not work within an instance group, but can be used to reorder an entire group (in reference to other top level pipeline/groups)

On the logistical side of things, we'll need to write the release notes and probably a docs PR for instance pipelines in general

aoldershaw commented:
Thanks for the thorough review @chenbh! Sorry again for the huge PR 😬
